bftengine/src/bftengine/ReplicaImp.cpp

```diff
@@ -673,6 +686,83 @@ bool ReplicaImp::relevantMsgForActiveView(const T *msg) {
   }
 }
 
+bool ReplicaImp::validatePreProcessedResultSignatures(preprocessor::PreProcessedResultMsg &msg) {
+  auto [buf, buf_len] = msg.getResultSignaturesBuf();
+  auto sigs = preprocessor::PreProcessResultSignature::deserializeResultSignatureList(buf, buf_len);
+
+  // f because the primary also executes the request (check PreProcessor::numOfRequiredReplies())
+  const auto expectedSignatureCount = getReplicaConfig().fVal;
+  if (sigs.size() < expectedSignatureCount) {
+    LOG_ERROR(GL,
+              "PreProcessedResult signatures validation failure - not enough signatures. Expected "
+                  << expectedSignatureCount << " got " << sigs.size());
+    return false;
+  }
+
+  for (const auto &s : sigs) {
+    auto hash = concord::util::SHA3_256().digest(msg.requestBuf(), msg.requestLength());
+    bool verificationResult = false;
+    if (getReplicaConfig().replicaId == s.sender_replica) {
+      std::vector<char> mySignature(sigManager_->getMySigLength(), 0);
+      sigManager_->sign((const char *)hash.data(), hash.size(), mySignature.data(), mySignature.size());
+      verificationResult = mySignature == s.signature;
```

Comment:
if replica himself gets his own message, then will there be any chance of changes? If no then we can just avoid checking the signature for same replica.